versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTransactionalDatabaseAdapter.java

```diff
@@ -832,63 +761,525 @@ protected Hash casOpLoop(
             this::tryLoopStateCompletion,
             config)) {
       while (true) {
-        GlobalStatePointer pointer = fetchGlobalPointer(ctx);
-        if (pointer == null) {
-          throw referenceNotFound(ref);
-        }
         Set<Hash> individualCommits = new HashSet<>();
         Set<Hash> individualKeyLists = new HashSet<>();
 
-        GlobalStatePointer newPointer =
-            casOp.apply(ctx, pointer, individualCommits::add, individualKeyLists::add);
-        if (newPointer.getGlobalId().equals(pointer.getGlobalId())) {
-          return tryState.success(branchHead(pointer, ref));
+        NamedReference refHead = fetchNamedReference(ctx, ref.getName());
+        if ((refHead == null || refHead.getRef().getType() != protoTypeForRef(ref))
+            && opVariant != CasOpVariant.CREATE_REF) {
+          // For all operations, except createReference, verify that the reference exists.
+          throw referenceNotFound(ref);
+        }
+        CasOpResult result =
+            casOp.apply(
+                ctx,
+                refHead != null ? refHead.getRef() : null,
+                individualCommits::add,
+                individualKeyLists::add);
+
+        boolean casSuccess;
+        switch (opVariant) {
+          case CREATE_REF:
+            casSuccess = createNamedReference(ctx, ref, result.newHead);
+            break;
+          case DELETE_REF:
+            casSuccess = deleteNamedReference(ctx, ref, refHead.getRef());
+            break;
+          case REF_UPDATE:
+          case COMMIT:
+            if (refHead.getRef().getHash().equals(result.newHead.asBytes())) {
+              // Target branch HEAD did not change
+              return tryState.success(result.newHead);
+            }
+            casSuccess = updateNamedReference(ctx, ref, refHead.getRef(), result.newHead);
+            break;
+          default:
+            throw new UnsupportedOperationException("Unknown opVariant " + opVariant);
         }
-        Hash branchHead = opVariant.deleteRef ? null : branchHead(newPointer, ref);
 
-        if (pointer.getGlobalId().equals(newPointer.getGlobalId())) {
-          throw hashCollisionDetected();
+        if (!casSuccess) {
+          if (opVariant == CasOpVariant.COMMIT) {
+            cleanUpCommitCas(ctx, individualCommits, individualKeyLists);
+          }
+
+          tryState.retry();
+          continue;
         }
 
-        if (globalPointerCas(ctx, pointer, newPointer)) {
-          return tryState.success(branchHead);
-        } else if (opVariant.commitOp) {
-          if (branchHead != null) {
-            individualCommits.add(branchHead);
+        // CAS against branch-heads succeeded, now try to write the ref-log-entry
+
+        int stripe = refLogStripeForName(ref.getName());
+        while (true) {
+          RefLogParents refLogParents = fetchRefLogParents(ctx, stripe);
+
+          RefLogEntry newRefLog = writeRefLogEntry(ctx, refLogParents, result.refLogEntry);
```

Comment:
I guess ref log entries may be reordered with respect to commits, e.g. server S1 makes commit `1` for ref `A`, the server S2 makes commit `2` for ref `A`, then server S2 makes ref log `A2`, then server S1 makes ref log `A1`. Is that a problem?

If not, I suppose we might as well have a flexible number of stripes. Even different servers could be using different numbers of stripes (as long as they know an upper bound for iteration)... and we could select a stripe randomly for each write. WDYT?

Reply:
Hrm - you're right - this can violate assumptions.

With global state, we guaranteed that the order in the ref-log matches the order in the commit-log for each individual branch. With this change, there's a chance that the order ref-log entries for commits on a branch does not match the order of commit-log entries.

I'm okay with violating the previous guarantee, as I do not see a legit use case for it. And that guarantee complicates things a lot. Thankfully we never documented that guarantee.

We do have some legit requirements like knowing the "lost HEADs" for GC (deleted references, re-assigned references) - but  that's mostly it.

I'd like to keep the split-calculation based on the reference name though. It's a best-effort approach though - but probably good enough.

WDYT?